Issue 60: Allow http://openrosa.org/xforms namespace for meta

mdudzinski opened the pull request:
This work is based on #68 but it uses `namespacePrefix` field instead of `namespacePrefixesByURI`.

dcbriccetti commented on src/org/javarosa/xform/parse/XFormParser.java:
```diff
@@ -330,15 +328,12 @@ public FormDef parse() throws IOException {
 
-    /** Extracts the namespaces from the html element, if present, and creates a map of URI to prefix */
-    private Map<String, String> buildNamespacesMap() {
+    /** Extracts the namespaces from the element, if present, and creates a map of URI to prefix */
```
The “if present, ” in the comment no longer applies.

mdudzinski replied:
@dcbriccetti that's right, thanks! I'll fix it.

dcbriccetti commented on test/org/javarosa/xform/parse/XFormParserTest.java:
```diff
@@ -49,6 +77,50 @@ public void parsesMetaNamespaceForm() throws IOException {
     }
 
+    @Test
+    public void serializeAndRestore_MetaNamespaceFormInstance() throws IOException {
```
Did you intend to use an “_”?

mdudzinski replied:
Yes because I like the idea of having underscore in tests names because it's easier to read them and those test methods are never called by hand.
However all existing tests use camelCase so perhaps we shouldn't mix the conventions. 
At least in a single test class. I'll change this. 
Thanks!

lognaturel commented:
Thank you both very much @dcbriccetti and @mdudzinski! 

@dcbriccetti and I did a live review together in #javarosa-code. The only thing I was left wondering about was whether `serializeAndRestoreMetaNamespaceFormInstance` should additionally show that the namespace in the form def and the namespace in the filled out instance match. If you think that would be a helpful extra check, it can be added in a separate PR. 

And now that I think about it, I have an additional question about the test. I noticed that XML generated for the filled out instance has the `orx2` prefix defined on the `data` node when I did `xml.writeExternal(new DataOutputStream(System.out));`. As @mdudzinski commented before, Collect filled out instances define the prefix on the `meta` node. What accounts for that difference?

mdudzinski commented:
@lognaturel @dcbriccetti Very good point! I somehow thought namespaces are being exercised under the hood but they are not. I added missing assertions in #155.

Yes the serialization of the namespaces is a bit tricky. In kxml2, the prefixes -> URI map is kept in `org.kxml2.kdom.Element.prefixes` field. There can be multiple namespaces definitions (prefix->URI maps), each defined at different tree level just like in an xml file. When `KXmlSerializer` is serializing a `Document` instance and finds an `Element` instance that has an URI in a `namespace` field that hasn't been defined in the element's its ancestors `prefixes` field, then serializer provides own prefix value for this namespace and puts the definition (in xml)  in the element's tag.

Calling `loadNamespaces(_xmldoc.getRootElement(), fi);` in `private void parseDoc(Map<String, String> namespacePrefixesByUri)` loads the namespaces defined in the form root element (html tag) into the form instance root element (data tag). This ensure that prefixes are consistent across restored form instance and the form definition but in the form instance xml file, namespaces definition is put in the root tag.

So for the test case example, without the above code, the result of the form instance serialization was:
<img width="378" alt="n0-namespace-prefix" src="https://user-images.githubusercontent.com/1262799/28415139-c121d76e-6d4e-11e7-8813-84a356818dc7.png">
with the above code:
<img width="674" alt="orx2-namespace-prefix" src="https://user-images.githubusercontent.com/1262799/28415566-a294da6a-6d50-11e7-9a38-ad369697434e.png">

There is one drawback though. This code will add all namespaces definitions to the form instance, even those not used by the instance. This can be fixed by filtering the prefixes map before attaching  it to the form instance. Let me know if that is a concern, I can implement this.

In general, the whole problem is caused by kxml2 implementation that doesn't handle namespace prefixes :(

I added the missing assertions and some comments for the above situation in #155.
